Fix specular edge artifacts

joewhatkins opened the pull request:
Update the specularPDF method so that it takes the absolute value of several parameters - see https://www.pbr-book.org/3ed-2018/Light_Transport_I_Surface_Reflection/Sampling_Reflection_Functions for an in-depth explanation.

This change fixes the issues with white highlights on glancing edges described in #215 

**Before**
![image](https://user-images.githubusercontent.com/85718573/180629834-59d4b308-bf60-46ce-ab0f-5afaa2e2b0da.png)

**After**
![image](https://user-images.githubusercontent.com/85718573/180629871-3bed036b-4428-47fe-a3ce-682cc7318895.png)

gkjohnson commented:
Thanks! Can you talk a bit about the change from the equations in the `jcgt` paper to the ones in pbr book? Should we change the direction sampling to align with the PDF function now, as well?

joewhatkins commented:
Actually, I couldn't parse out how equation 27 in http://jcgt.org/published/0003/02/03/ mapped to the formula I corrected from #174 - nor could I parse out how equation 17 did (the cited equation before the change in #174).

In the process of trying to figure all of this out, I read through http://jcgt.org/published/0007/04/01/, a different paper hosted at the same site, that is cited in shaderGGXFunctions.js. Equation 17 in http://jcgt.org/published/0007/04/01/ does seem to correspond to specularPDF - so I think it's possible that someone meant to cite http://jcgt.org/published/0007/04/01/ but cited http://jcgt.org/published/0003/02/03/ instead.

As far as I can tell, [e4e2757](https://github.com/gkjohnson/three-gpu-pathtracer/pull/174/commits/e4e275746877aea2579bc6a29792a0137da5bb70) did fix a legitimate error - the specular PDF should be computed from the outgoing direction - but we should be using wo in the denominator instead of wi.

The equations I highlighted in the PBR book are very similar to the ones outlined in http://jcgt.org/published/0007/04/01/ but include some pseudocode with some differences that fixed the issues with highlights when I incorporated them.

The other changes involved taking the absolute value of some of the normal and dot product terms in the ggxPDF method. I broke this out from the ggxPDF method in shaderGGXFunctions.js since I figured that might affect things other than these specular problems. That might be a thing to incorporate into a bigger PR. Either way this seems to fix the issue.

All of that said, I'd love to hear more about [e4e2757](https://github.com/gkjohnson/three-gpu-pathtracer/pull/174/commits/e4e275746877aea2579bc6a29792a0137da5bb70) from @mrxz to make sure my understanding is correct.

gkjohnson commented:
Okay great, thanks for the explanation! It would have been me that messed up the comment references. It's on my list to revisit the GGX and sampling functions to address some other issues, too, so I'll take another pass at the comments, as well. This is definitely an improvement, though, so I'll merge this before the next release!

mrxz commented:
Since we compute the `halfVector` from `wi` and `wo` the dot product of `wi` and `halfVector` is the same as the dot product of `wo` and `halfVector`. So it should be fine to use either in the denominator. 

The thing I don't see yet is which change actually solved the issue and why. I would expect `dot( wo, halfVector )` to be always positive in the specularPdf case. Otherwise, shouldn't we also take the abs of it in the denominator?
